### ISSUE TITLE
Downstream headers missing or incorrect in pipelines with got.stream

### ISSUE THREAD

**Issue by retrocausal** [sindresorhus/got#1433]
#### What would you like to discuss?
The migration from request guide states 

> // When someone makes a request to our server, we receive a body and some headers.
		// These are passed to Got. Got proxies downloaded data to our server response,
		// so you don't have to do `response.writeHead(statusCode, headers)` and `response.end(body)`.
		// It's done automatically.

After reading other bits of documentation on how to stream using `got` , I have now tried both 
`stream.pipeline` and `req.pipe(got(...)).pipe(res)`

my options object passed to `got` , has `decompress: false` as the request sent from the browser already sends the `accept-encoding` header

When I use the now deprecated `request` library to pipe from the stream to the  express server response 
`req.pipe(request({...})).pipe(res)` , all of the response headers set by the remote server are set downstream on the `res` as well and the file successfully downloads on the browser. 

However, when I replace `request` with `got.stream` , the `content-encoding`  header is missing downstream after the server-response reaches the browser, and the header `content-length` reads `0`

**headers set by the remote server that is queried for the file:**

```
 content-disposition: attachment; filename=statement.csv 
 content-encoding: gzip 
 content-length: 568 
 content-type: application/octet-stream 
 date: Wed, 02 Sep 2020 16:46:12 GMT 
 vary: Accept-Encoding 
```
**headers received downstream via `Request`** : 

<img width="1348" alt="Screenshot 2020-09-02 at 10 32 20 PM" src="https://user-images.githubusercontent.com/15086525/92014041-3cc4c200-ed6c-11ea-8305-463901af9454.png">

**headers received downstream via `Got`**:

<img width="1350" alt="Screenshot 2020-09-02 at 10 28 32 PM" src="https://user-images.githubusercontent.com/15086525/92014111-582fcd00-ed6c-11ea-896e-9ffca7300233.png">

**Note:**  This behaviour is irrespective of whether I use `pipelines` or `await pipelines` or use the vanilla `pipe` sequences with custom error handlers

I want to know If there is something I am missing on the usage front, or if there are any workarounds if it is a known issue

My code (now with the pipe from to) : 

```js
request
          .pipe(got.stream(source, options))
          .on("error", (error) => {
            const rejection = this.exceptionHandler(error, false);
            reject(rejection);
          })
          .pipe(response)
          .on("error", (error) => {
            const rejection = this.exceptionHandler(error, false);
            reject(rejection);
          })
          .on("finish", () => {
            resolve();
          });
```

and the options passed to `got` :

```
{
  retry: 0,
  headers: {
    'accept-language': 'en-GB,en-US;q=0.9,en;q=0.8,hi;q=0.7',
    'accept-encoding': 'gzip, deflate, br',
    referer: 'http://localhost:3000/console/reports?page=bank_statement',
    'sec-fetch-dest': 'iframe',
    'sec-fetch-user': '?1',
    'sec-fetch-mode': 'navigate',
    'sec-fetch-site': 'same-origin',
    accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
    'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.83 Safari/537.36',
    dnt: '1',
    'upgrade-insecure-requests': '1',
    'cache-control': 'no-cache',
    pragma: 'no-cache',
    connection: 'close',
    host: 'localhost:9009',
    Authorization: 'bearer <token>',
    'X-Client-Id': 'abc'
  },
  searchParams: {
    XCI: 'abc',
    from: '1596635338860',
    to: '1599140938860',
    bank: 'xyz'
  },
  method: 'get',
  decompress: false
}
```
Node: v12.x / v14.9.0
OS: Mojave MacOS
got: 11.5.1/2 /11.6.0

 
...

#### Checklist

- [X] I have read the documentation.

**Comment by szmarczak** [comment]
Can you create a RunKit example please?

**Comment by retrocausal** [comment]
Does [This runkit](https://runkit.com/retrocausal/runkit-npm-express) look sane to you? 
this is the first time I have created a zlib compression on the fly. there might be some amateurish  stuff

**Comment by szmarczak** [comment]
> my options object passed to got , has decompress: false

Well I can clearly see that you don't in the runkit example.

**Comment by retrocausal** [comment]
> > my options object passed to got , has decompress: false
> 
> Well I can clearly see that you don't in the runkit example.

Nope I do have it set to false 

<img width="527" alt="Screenshot 2020-09-05 at 10 26 09 AM" src="https://user-images.githubusercontent.com/15086525/92297987-4748ac80-ef62-11ea-8627-60ccede10c72.png">

Few things to note : 

the runkit server negotiates with `http2` and not `http1.1` . I dont really know if having a different protocol makes a difference, but the `connection : close `  header is not present in h2 responses. 

I feel in the project, the response is not fully read from source before piping it back to the browser, or if an error occurs because of premature connection loss, that is not relayed back to be handled. 

Is there a way to inspect chunks received from the API endpoint before piping it to the response? 

I tried `.on(response, response)` , but that does not give me the body / chunk that is being streamed

**Comment by szmarczak** [comment]
> Nope I do have it set to false

This piece of code is missing in your RunKit example. Please either update your RunKit example or create a reproducible repo.

**Comment by retrocausal** [comment]
@szmarczak . I dont really know why runkit is showing us both different versions of master. 
I seem to have a break through though, seems to be a case of  remote path (url construct) and search params being parsed or overridden (improper library usage I guess )

I will dig more and might have some queries on how to plug in got properly. 

Would it be fine If I come back with usage queries and clarifications? or should I fill out a new discussion template?

**Comment by szmarczak** [comment]
Can you just create a repo to reproduce the bug? That would work best.

**Comment by retrocausal** [comment]
> Can you just create a repo to reproduce the bug? That would work best.

Hey I did not mean to open this as a bug, I remember filling a discussion template :) 
Mostly because I wasn't sure I was integrating `got` right . 

So, after comparing the responses I got from both `request` and `got` , I found that the query params sent to the remote server differed. 

**Reason :** 

1.  I had used `?date_from=value1&date_to=value2` in the path for `got` options
2. further, I had also added `searchParams` to `got` 
 > 
```js
       searchParams: {
          XCI: 'abc',
          from: '1596635338860',
          to: '1599140938860',
          bank: 'xyz'
     }
```
These params were taken from the `clientRequest` object

and I completely missed [this section of the readme](https://github.com/sindresorhus/got#searchparams)  where it says 

> Query string that will be added to the request URL. This will override the query string in url


so prioritising what I needed to be passed, fixed the issue. 

**Actual Issue: **

When the overridden params were passed to the server, it did not consider the request to be proper. hence did not attach the 
`content-encoding` and `length` headers. In hindsight, a proper error or a `bad request` from the server would have saved us this conversation
